pkg/controller/render/render_controller.go

```diff
@@ -486,7 +486,12 @@ func (ctrl *Controller) syncGeneratedMachineConfig(pool *mcfgv1.MachineConfigPoo
 		return err
 	}
 
-	generated, err := generateRenderedMachineConfig(pool, configs, cc)
+	currentMC, err := ctrl.mcLister.Get(pool.Spec.Configuration.Name)
```

Comment:
Just as an initial comment, spec.config is the current rendered config of the pool, but that doesn't mean there isn't e.g. an existing update or failing nodes on different versions.

Instead of getting spec here, should we either process status, or loop through all nodes and check for the currentConfig annotation to see whether this is reconcileable?

The reason why we had it in the daemon in the first place is because each individual node might be on different versions. Although this isn't the case most of the time, I'm concerned that we could leave some nodes out of the calculation.

Reply:
I don't think RenderController takes each individual node into account, does it? To be clear, I'm not removing the call from the daemon. Instead, I relocated this code with the intent that both RenderController and the daemon could use it. The daemon path however, will do some additional checks because it has access to the nodes' filesystem (which we don't here in the RenderController). It's kind of a "Trust But Verify" type of situation.

An alternative approach would be to additionally relocate `generateRenderedMachineConfig()` into `pkg/controller/common`. Doing this would allow us to reuse that code here as well as in the BuildController to do a preemptive check before performing the on-cluster build.

Reply:
Ah, having multiple checks works, I was only thinking from the render controller's perspective, but it makes more sense that the controller has an initial safety check as well :+1:

Reply:
I'll briefly mention that I tried to do what I suggested about moving the generation code into `pkg/controller/common` and then tried to add that into the BuildController, but I really didn't like how that turned out. I think this pattern might ultimately be what makes the most sense for right now.